ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/BackingSlots.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2020, Red Hat, Inc. and/or its affiliates,
+ * and individual contributors as indicated by the @author tags.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2020,
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com)
+ */
+package com.arjuna.ats.internal.arjuna.objectstore.slot;
+
+import java.io.IOException;
+
+/**
+ * Interface for pluggable internal implementations of the SlotStore backend.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-04
+ */
+public interface BackingSlots {
+
+    /**
+     * Initialize the instance from the given configuration.
+     * This must be called once per instance, before any other method is called.
+     *
+     * @param config The configuration to apply.
+     */
+    void init(SlotStoreEnvironmentBean config) throws IOException;
+
+    /**
+     * Update the given slot with the provided data, overwriting (non-atomically) any existing data.
+     * The update may not be immediately persistent, depending on config syncWrites
+     *
+     * @param slot the index, from 0 to config numberOfSlots-1
+     * @param sync true for immediate persistence, false otherwise
+     * @param data the content.
+     */
+    void write(int slot, byte[] data, boolean sync) throws IOException;
+
+    /**
+     * Read the given slot, returning its contents.
+     *
+     * @param slot the index, from 0 to config numberOfSlots-1
+     * @return the content, or null if the slot has not been written or has been cleared.
```

Comment:
yes, though I'm treating the checksum as an internal implementation detail. It's really validating the write completed and didn't get interrupted by a crash. We could do that a different way without changing the API, so it's not documented behaviour. Broken checksum is just a way of determining the slot has not been written (completely), which is documented API behaviour.

Reply:
So a corrupt slot is equivalent to one that had never been written to in the first place. What if it is written okay, read and updated but the second write corrupted it. The read may have real world consequences (e.g. to tell an XAResource to do something). Since there is no log message (that could usefully be used to correlate with other log messages) about the corrupt slot the information (that the slot was corrupt) is lost.

Reply:
This is really covered below. I'll copy here:

A partial write is never exposed, it just behaves as never having happened, so the slot is considered to still be empty. There is nothing useful a user can do about it even if it's made visible to them, so they don't need to know. The mechanism that puts overwrites for the same key into parallel slots is critical here, as if we let data vanish by partially overwriting it then crashing, that would be a user visible data loss that would not be recoverable.

Logically, you may have a partial update, but also the entire previous state. Recovering that is essentially an internal rollback - the partial update in one slot is considered invalid, freeing the slot, and the original state is restored from the other slot. The index update occurs only after a sucessfull write, so a partially written state can't be seen externally.

Reply:
The scenario I quoted below was:

> ... the parallel new slot can never become corrupt after it has been written

Reply:
So this conversation was resolved by a [later comment](https://github.com/jbosstm/narayana/pull/1593#issuecomment-615253638).